kong/hooks.lua

```diff
@@ -0,0 +1,75 @@
+local _M = {}
+
+
+local hooks = {}
+
+
+local ipairs = ipairs
+local pack = table.pack
+local unpack = table.unpack
+local insert = table.insert
+
+
+local function wrap_hook(f)
+   return function(acc, ...)
+      if acc and not acc[1] then
+         return acc
+      end
+      return pack(f(...))
+   end
+end
+
+
+function _M.register_hook(name, hook, opts)
```

Comment:
The name `raw` seems very ... "implementation oriented": it is "raw" because it is "not wrapped" - but the fact that we are wrapping it into something else is an implementation detail. What the user of this function cares about is whether this hook will execute when there are errors in other hooks before it.

I therefore recommend renaming the option to `ignore_errors` or even `ignore_previous_errors`.

Reply:
Given this has been agreed upon before with @hishamhm and @kidd, I'd also like to hear their thoughts.

Reply:
@kikito Understood, but that was not the rationale.

`raw` here was meant to imply that the function needs to follow the low-level protocol for hook functions when it cares about controlling the chaining sequence. The fact that the default chaining sequence behavior is to abort on errors is meant to be the implementation detail.

What raw functions allow is a lot more than ignoring previous errors. You could for example take the output from the previous hook and combine it to the result of the current hook in arbitrary ways. (In other words, you can also implement map, reduce, etc. — in other other words, what the hooks system implements is a fold, which is a super general construct for a super general feature).

The documentation there (mea culpa, I wrote the docs bit) is misleading by focusing on the "handling errors" example. Perhaps it should have even left out the default chaining behavior altogether and leave it unspecified (the more I think about it, the more I think it should, because you should only use the not-raw ones if you _don't care_ about the chaining sequence). If all the `raw` flag did was `ignore_previous_errors`, we wouldn't need the general system.

Perhaps we could can another name like `low_level` to avoid the "raw vs. wrapped" comparison — that's good feedback, thanks for bringing it up!

Reply:
Thanks for the comment @hishamhm. @kikito and I just met and made two changes:
- Renamed the flag to `low_level`
- Adjusted the docs to not focus on the error handling part